client/rpc-servers/src/lib.rs

```diff
@@ -39,194 +40,119 @@ pub const WS_MAX_BUFFER_CAPACITY_DEFAULT: usize = 16 * MEGABYTE;
-	tokio_handle: tokio::runtime::Handle,
-) -> io::Result<http::Server> {
-	let max_request_body_size = maybe_max_payload_mb
+	max_payload_mb: Option<usize>,
+	metrics: Option<RpcMetrics>,
+	rpc_api: RpcModule<M>,
+	rt: tokio::runtime::Handle,
+) -> Result<HttpServerHandle, anyhow::Error> {
+	let max_request_body_size = max_payload_mb
 		.map(|mb| mb.saturating_mul(MEGABYTE))
 		.unwrap_or(RPC_MAX_PAYLOAD_DEFAULT);
 
-	http::ServerBuilder::new(io)
-		.threads(1)
-		.event_loop_executor(tokio_handle)
-		.health_api(("/health", "system_health"))
-		.allowed_hosts(hosts_filtering(cors.is_some()))
-		.rest_api(if cors.is_some() { http::RestApi::Secure } else { http::RestApi::Unsecure })
-		.cors(map_cors::<http::AccessControlAllowOrigin>(cors))
-		.max_request_body_size(max_request_body_size)
-		.start_http(addr)
-}
+	let mut acl = AccessControlBuilder::new();
+
+	if let Some(cors) = cors {
+		// Whitelist listening address.
+		// NOTE: set_allowed_hosts will whitelist both ports but only one will used.
+		acl = acl.set_allowed_hosts(format_allowed_hosts(addrs))?;
+		acl = acl.set_allowed_origins(cors)?;
+	};
+
+	let builder = HttpServerBuilder::new()
+		.max_request_body_size(max_request_body_size as u32)
+		.set_access_control(acl.build())
+		.custom_tokio_runtime(rt.clone());
+
+	let rpc_api = build_rpc_api(rpc_api);
+	let handle = if let Some(metrics) = metrics {
+		let middleware = RpcMiddleware::new(metrics, "http".into());
+		let builder = builder.set_middleware(middleware);
+		let server = tokio::task::block_in_place(|| rt.block_on(async { builder.build(addrs) }))?;
+		server.start(rpc_api)?
+	} else {
+		let server = tokio::task::block_in_place(|| rt.block_on(async { builder.build(addrs) }))?;
+		server.start(rpc_api)?
+	};
 
-/// Start IPC server listening on given path.
-pub fn start_ipc<M: pubsub::PubSubMetadata + Default>(
-	addr: &str,
-	io: RpcHandler<M>,
-	server_metrics: ServerMetrics,
-) -> io::Result<ipc::Server> {
-	let builder = ipc::ServerBuilder::new(io);
-	#[cfg(target_os = "unix")]
-	builder.set_security_attributes({
-		let security_attributes = ipc::SecurityAttributes::empty();
-		security_attributes.set_mode(0o600)?;
-		security_attributes
-	});
-	builder.session_stats(server_metrics).start(addr)
+	log::info!("Starting JSON-RPC HTTP server: addr={:?}, allowed origins={:?}", addrs, cors);
+	Ok(handle)
 }
 
 /// Start WS server listening on given address.
-pub fn start_ws<
-	M: pubsub::PubSubMetadata + From<futures::channel::mpsc::UnboundedSender<String>>,
->(
-	addr: &std::net::SocketAddr,
+pub fn start_ws<M: Send + Sync + 'static>(
+	addrs: &[SocketAddr],
 	max_connections: Option<usize>,
 	cors: Option<&Vec<String>>,
-	io: RpcHandler<M>,
-	maybe_max_payload_mb: Option<usize>,
-	maybe_max_out_buffer_capacity_mb: Option<usize>,
-	server_metrics: ServerMetrics,
-	tokio_handle: tokio::runtime::Handle,
-) -> io::Result<ws::Server> {
-	let max_payload = maybe_max_payload_mb
+	max_payload_mb: Option<usize>,
```

Comment:
we need to add a parameter here to make it possible to provide a custom `subscriptionID generation` such as ethereum requires, https://github.com/paritytech/frontier/pull/534